File: examples/ts/stateful/Stateful.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'Stateful.runar.ts'), 'utf8');

describe('Stateful', () => {
  it('initialises count and maxCount', () => {
    const c = TestContract.fromSource(source, { count: 0n, maxCount: 10n });
    expect(c.state.count).toBe(0n);
    expect(c.state.maxCount).toBe(10n);
  });

  it('increments up to the maximum', () => {
    const c = TestContract.fromSource(source, { count: 0n, maxCount: 10n });
    expect(c.call('increment', { amount: 3n }).success).toBe(true);
    expect(c.state.count).toBe(3n);
    expect(c.call('increment', { amount: 7n }).success).toBe(true);
    expect(c.state.count).toBe(10n);
  });

  it('rejects increments that exceed maxCount', () => {
    const c = TestContract.fromSource(source, { count: 0n, maxCount: 10n });
    const result = c.call('increment', { amount: 11n });
    expect(result.success).toBe(false);
  });

  it('reset clears the count back to zero', () => {
    const c = TestContract.fromSource(source, { count: 5n, maxCount: 10n });
    expect(c.call('reset').success).toBe(true);
    expect(c.state.count).toBe(0n);
  });
});

File: examples/zig/stateful/Stateful_test.zig
const std = @import("std");
const root = @import("../examples_test.zig");
const Stateful = @import("Stateful.runar.zig").Stateful;

fn contractPath(comptime basename: []const u8) []const u8 {
    return "stateful/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check Stateful.runar.zig" {
    try runCompileChecks("Stateful.runar.zig");
}

test "Stateful.increment advances count up to the maximum" {
    var c = Stateful.init(0, 10);
    c.increment(3);
    try std.testing.expectEqual(@as(i64, 3), c.count);
    c.increment(7);
    try std.testing.expectEqual(@as(i64, 10), c.count);
}

test "Stateful.reset clears count back to zero" {
    var c = Stateful.init(5, 10);
    c.reset();
    try std.testing.expectEqual(@as(i64, 0), c.count);
}
